src/app/jscaip/AI/MCTSWithHeuristic.ts

```diff
@@ -0,0 +1,65 @@
+import { MGPOptional, Utils } from '@everyboard/lib';
+import { GameStatus } from '../GameStatus';
+import { Move } from '../Move';
+import { Player } from '../Player';
+import { SuperRules } from '../Rules';
+import { EmptyRulesConfig, RulesConfig } from '../RulesConfigUtil';
+import { GameState } from '../state/GameState';
+import { MoveGenerator } from './AI';
+import { BoardValue } from './BoardValue';
+import { GameNode } from './GameNode';
+import { MCTS } from './MCTS';
+import { Heuristic, HeuristicBounds } from './Minimax';
+
+/**
+ * Like MCTS, but uses a heuristic function to evaluate non-terminated states.
+ */
+export class MCTSWithHeuristic<M extends Move,
+                               S extends GameState,
+                               C extends RulesConfig = EmptyRulesConfig,
+                               B extends BoardValue = BoardValue,
+                               L = void>
+    extends MCTS<M, S, C, L>
+{
+
+    public constructor(name: string,
+                       moveGenerator: MoveGenerator<M, S, C>,
+                       rules: SuperRules<M, S, C, L>,
+                       private readonly heuristic: Heuristic<M, S, B, C>)
+    {
+        super(name, moveGenerator, rules);
+        // this.maxGameLength = 10; // no need to go further since we have a heuristic
+    }
+
+    /**
+     * Return a win score which is the average of all metrics
+     */
+    protected override winScore(node: GameNode<M, S>,
+                                config: MGPOptional<C>,
+                                gameStatus: GameStatus,
+                                player: Player): number {
+        if (gameStatus === GameStatus.ONGOING) {
+            const boardValue: B = this.heuristic.getBoardValue(node, config);
+            const optionalBounds: MGPOptional<HeuristicBounds<B>> = this.heuristic.getBounds(config);
+            Utils.assert(optionalBounds.isPresent(),
+                         'MCTSWithHeuristic used with a heuristic that has no max value, please define getBounds');
+            const bounds: HeuristicBounds<B> = optionalBounds.get();
+            Utils.assert(boardValue.metrics.length === bounds.player0Max.metrics.length &&
+                         boardValue.metrics.length === bounds.player1Max.metrics.length,
+                         'Metrics and bound values should have the same shape');
+            let value: number = 0;
+            for (let i: number = 0; i < boardValue.metrics.length; i++) {
+                Utils.assert(bounds.player0Max.metrics[i] <= boardValue.metrics[i] &&
+                             boardValue.metrics[i] <= bounds.player1Max.metrics[i],
```

Comment:
c'est spécial comme logique quand même, ça devrais pas être simplement entre le minimum pour ce joueur et le maximu pour ce joueur  ?

Reply:
la métrique i doit être entre le min (aka le max de joueur 0) et le max (aka le max de joueur 1)
Car pour rappel la métrique peut mélanger les joueurs (faire j1 - j0)

Reply:
là on a que des métriques best ou on fait of(0, 48) et (48, 0) ou un truc symétrique répétitif du genre.
On pourrais avoir une proof en contexte quand le meilleurs score d'un joueur est pas le même,
Genre le Brandhub le "score" étant le nombre de pièces, et ce nombre étant asymétrique, l'implémenter serait pas pire ainsi on voti un of(5, 8) / of (8, 5) ou j'sais pas ?

Reply:
c'est intéressant comme ticket prochain, mais pas dans le cadre de celui-ci :p

Reply:
ah not agreeing, c'est un proof of concept de ton code actuel en place !
Eat your own dog food !

Reply:
Ouais le Tafl n'est pas du tout amélioré haha, bah du coup, why am I reviewing this :D ?